### PR TITLE
[Core] Making kratos applications unloadable

### PR DESCRIPTION
**Pull request by roigcarlo**
**📝 Description**
This PR allows `KratosApplication` class to be able to unload the components that has registered in the KratosCommponents containers during its unload.

The register of the component source is done automatically with the name of the application. If the register is done outside the KratosApplication class, I've provided a macro to manually define the name of the source.

This is needed by #11506 in order to avoid having the kernel unloading everything.

Its very likely that I've missed something so comments are appreciated.

@rfaasse @avdg81 I would like to merge this before the tests as this is a much more clean way to unload the kernel than what we have in #11506, and also will reduce its size and complexity.

**🆕 Changelog**
- Added the possibility for KratosApplication to unload its components
- Added a new KratosComponentsSource to keep track of who has added which components to each component containers
- KratosApplication now unloads its components when destroyed
- KratosApplication Base class now unloads LinearSolvers and Preconditioners custom Components.
- Changing some namings of Spaces to be consistent and avoiding missues form other parts of the code
- `AdditiveSchwarzPreconditioner` is now registered as part of the application instead of being part of the interface.

### CODE REVIEW

**Comment by roigcarlo** [comment]
@philbucher is https://github.com/KratosMultiphysics/Kratos/pull/12281/commits/eae5b7fd1d81d797f7499dd57e774abecbf59599 correct?

File: kratos/factories/linear_solver_factory.h [hunk 1]
```diff
@@ -151,31 +151,33 @@ inline std::ostream& operator << (std::ostream& rOStream,
 }
 ///@}
 
-typedef TUblasSparseSpace<double> SparseSpaceType;
-typedef TUblasDenseSpace<double> LocalSparseSpaceType;
+typedef TUblasSparseSpace<double> SpaceType;
```
Comment (loumalouomega) [inline kratos/factories/linear_solver_factory.h:154]: Why this renaming?

File: kratos/factories/preconditioner_factory.h [hunk 2]
```diff
@@ -150,18 +150,19 @@ inline std::ostream& operator << (std::ostream& rOStream,
 
 ///@}
 
-typedef TUblasSparseSpace<double> SparseSpaceType;
-typedef TUblasDenseSpace<double> LocalSparseSpaceType;
+typedef TUblasSparseSpace<double> SpaceType;
```
Comment (loumalouomega) [inline kratos/factories/preconditioner_factory.h:153]: Why renaming?

File: kratos/factories/standard_linear_solver_factory.cpp [hunk 3]
```diff
@@ -64,8 +64,9 @@ namespace Kratos
         static auto MonotonicityPreservingSolverFactory= StandardLinearSolverFactory<SpaceType,LocalSpaceType,MonotonicityPreservingSolverType>();
         static auto SkylineLUComplexSolverFactory = StandardLinearSolverFactory<ComplexSpaceType, ComplexLocalSpaceType, SkylineLUComplexSolverType>();
 
-        //registration of linear solvers
-//         KRATOS_REGISTER_LINEAR_SOLVER("LinearSolver", StandardLinearSolverFactory<SpaceType,LocalSpaceType,LinearSolverType>());
+        // Registration of linear solvers
+        KRATOS_SET_REGISTER_SOURCE("KratosMultiphysics");
```
Comment (loumalouomega) [inline kratos/factories/standard_linear_solver_factory.cpp]: Please move the comment, it is disleading

**Review by loumalouomega — COMMENTED** [review]
Some comments

File: kratos/factories/linear_solver_factory.h (hunk 1 above)
Comment (roigcarlo) [reply kratos/factories/linear_solver_factory.h:154]: Because in half ok the linear_solver_factory `TUblasSparseSpace` was refered as `SpaceType` and on the other half as `SparceSpaceType`.

Since the same happened to the `LocalSparseSpaceType`, and the name is wrong because should be `LocalDenseSpaceType`, I just removed the Sparse and Dense from the aliases and keep it consistent with the others aliases (for example the ones in the `standard_linear_solvers_factories.h`)

File: kratos/factories/preconditioner_factory.h (hunk 2 above)
Comment (roigcarlo) [reply kratos/factories/preconditioner_factory.h:153]: Same

File: kratos/factories/standard_linear_solver_factory.cpp (hunk 3 above)
Comment (roigcarlo) [reply kratos/factories/standard_linear_solver_factory.cpp]: Ok :+1:

File: kratos/factories/linear_solver_factory.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/factories/linear_solver_factory.h:154]: Puff, I think we need to clarify notation and then unify.

Comment (roigcarlo) [reply kratos/factories/linear_solver_factory.h:154]: Agree, but clearly defining as `Sparse` something that is `Dense` was wrong

**Review by matekelemen — COMMENTED** [review]
`Unregister*` => `Deregister*`

*Unregistered* means that something has never been registered, while *deregister* means that smth is in the registry and we want to remove it from there.

**Comment by roigcarlo** [comment]
Ping @KratosMultiphysics/technical-committee @pooyan-dadvand pretty please :)

File: applications/IgaApplication/iga_application.cpp [hunk 4]
```diff
@@ -59,6 +64,8 @@ KRATOS_INFO("") << "    KRATOS  _____ _____\n"
                 << "           |_____\\_____/_/    \\_\\\n"
                 << "Initializing KratosIgaApplication..." << std::endl;
 
+    RegisterPreconditioners();
```
Comment (philbucher) [inline applications/IgaApplication/iga_application.cpp:67]: that makes more sense, dont remember why we had it in the python interface first. I do remember thought that this stuff was sensitive, please double check that it still works

File: kratos/factories/standard_preconditioner_factory.cpp [hunk 5]
```diff
@@ -43,11 +43,24 @@ namespace Kratos
         static auto ILU0PreconditionerFactory= StandardPreconditionerFactory<SpaceType,LocalSpaceType,ILU0PreconditionerType>();
         static auto ILUPreconditionerFactory= StandardPreconditionerFactory<SpaceType,LocalSpaceType,ILUPreconditionerType>();
 
+        // Setting default register name
+        KRATOS_SET_REGISTER_SOURCE("KratosMultiphysics");
```
Comment (philbucher) [inline kratos/factories/standard_preconditioner_factory.cpp:47]: why is this needed? The kernel does this already

File: kratos/includes/define.h [hunk 6]
```diff
@@ -693,46 +693,58 @@ catch(...) { Block KRATOS_THROW_ERROR(std::runtime_error, "Unknown error", MoreI
...
+#endif
+#define KRATOS_SET_REGISTER_SOURCE(name) \
+    std::string mApplicationName = name;
+
 #ifdef KRATOS_REGISTER_GEOMETRY
 #undef KRATOS_REGISTER_GEOMETRY
 #endif
 #define KRATOS_REGISTER_GEOMETRY(name, reference) \
     KratosComponents<Geometry<Node>>::Add(name, reference); \
+    KratosComponents<Geometry<Node>>::AddSource(mApplicationName, name); \
```
Comment (philbucher) [inline kratos/includes/define.h:707]: thats rather dirty :sweat_smile: 
especially for the core

I cannot think of an easy solution without changing the interface, but please comment this properly. This can lead to nasty errors if used outside the application context

How about moving these macros to `kratos/includes/kratos_application.h`, then this would be much cleaner, and should not break existing code. (And reduce the size of `define.h` which I strongly advocate for)
What do you think?

File: kratos/includes/kratos_components.h [hunk 7]
```diff
@@ -93,6 +96,25 @@ class KRATOS_API(KRATOS_CORE) KratosComponents
...
+        // Create the container if its the first component source added
+        if (msComponentsSources.find(rSrcName) == msComponentsSources.end()) {
+            msComponentsSources[rSrcName] = std::unordered_set<std::string>();
+        }
+
+        // Check if a different object was already registered with this name
+        auto it_comp =  msComponentsSources.find(rCmpName);
+        KRATOS_ERROR_IF(it_comp != msComponentsSources.end()) << "An object with name \"" << rCmpName << " has already been registered for " << rSrcName << std::endl;
+        msComponentsSources[rSrcName].insert(rCmpName);
+    }
```
Comment (philbucher) [inline kratos/includes/kratos_components.h]: this function searches 3 times for `rSrcName` in `msComponentsSources`

Can you please improve this?
E.g. sth like
```c++
auto it_found = msComponentsSources.find(rSrcName);
if (it_found == msComponentsSources.end()) {
    it_found = msComponentsSources.insert(rSrcName, std::unordered_set<std::string>()); // probably wrong function, needs the one that returns the new iterator
}

// use it_found ...

```

File: kratos/includes/kratos_components.h [hunk 8]
```diff
@@ -104,6 +126,23 @@ class KRATOS_API(KRATOS_CORE) KratosComponents
...
     }
 
+    /**
+     * @brief Removes all component registered by the specified source and then removes the source itself.
+     * @param rSrcName The name of the component source to remove components from.
+     */
+    static void CleanKratosComponents(const std::string& rSrcName)
+    {
+        // If the component is not loaded, we don't try to unload it ( this prevents applications with the unregister function not implemented from crashing the core)
+        // if (msComponentsSources.find(rSrcName) != msComponentsSources.end()) {
```
Comment (philbucher) [inline kratos/includes/kratos_components.h]: please remove commented code, this is confusing

File: kratos/sources/kratos_application.cpp [hunk 9]
```diff
@@ -295,4 +295,22 @@ void KratosApplication::RegisterKratosCore() {
     // Register ConstitutiveLaw BaseClass
     KRATOS_REGISTER_CONSTITUTIVE_LAW("ConstitutiveLaw", mConstitutiveLaw);
 }
+
+void KratosApplication::DeregisterCommonComponents() 
+{
+    std::cout << "Deregistering " << mApplicationName << std::endl;
```
Comment (philbucher) [inline kratos/sources/kratos_application.cpp]: LOGGER_DETAIL or LOGGER_TRACE

**Review by philbucher — COMMENTED** [review]
some comments, otherwise looks good to me!

File: kratos/includes/define.h (hunk 6 above)
Comment (philbucher) [reply kratos/includes/define.h:707]: On second thought, I would even go further and not only move this, but also replace the macros with member functions of `KratosApplication`. This way it cannot be used wrong

This can be done in a separate PR ofc, but just mentioning what I would imagine the final solution for this to be

File: kratos/factories/standard_preconditioner_factory.cpp (hunk 5 above)
Comment (roigcarlo) [reply kratos/factories/standard_preconditioner_factory.cpp:47]: The new macro I introduced assumes that you are going to register things inside `KratosApplication`class (which is 99% of the case), hence `mApplicationName` exists.

For preconditioners and linear solvers in the core, and possibly others, registering is done outside the application class itself, and I need a way to track down where it belongs.

I will mark this as solved as is basically an extension of the comment below so we keep it centralized

File: kratos/includes/define.h (hunk 6 above)
Comment (roigcarlo) [reply kratos/includes/define.h:707]: Mostly agree, not the cleanest solution out there, but hard to make it better without changing the interface.

What bothers me the most is having  `mApplicationName` as an argument hidden to the user. Moving the macro definitions to `kratos_application.h` will alleviate some possible misuses, but still has its drawbacks (the register of LinearSolvers, Preconditioners and pretty much everything that register components outside the `Register` function will break, as those do not dependent on the application, which is what I commented above and the reason to intruduce the `KRATOS_SET_REGISTER_SOURCE`)

About moving them as functions of `KratosApplications` I disagree. If we do that we have to expose every possible instantiation of variable, which will prohibit making use of KratosComponents of arbitrary types

File: applications/IgaApplication/iga_application.cpp (hunk 4 above)
Comment (roigcarlo) [reply applications/IgaApplication/iga_application.cpp:67]: Everything passing unless I've overseen it catastrophically, seems more to me like it was added before any example of preconditioned was present into the core.

I will take a second look just in case.

File: kratos/includes/kratos_components.h (hunk 7 above)
Comment (roigcarlo) [reply kratos/includes/kratos_components.h]: Oh, wait there is a mistake, the second find should be done over:

```c++
// Check if a different object was already registered with this name
auto it_comp =  msComponentsSources[rSrcName].find(rCmpName);
KRATOS_ERROR_IF(it_comp != msComponentsSources[rSrcName].end()) << "An object with name \"" << rCmpName << " has already been registered for " << rSrcName << std::endl;
msComponentsSources[rSrcName].insert(rCmpName);
```

Does this make more sense?

File: kratos/includes/define.h (hunk 6 above)
Comment (philbucher) [reply kratos/includes/define.h:707]: I meant to move only the macros for the types that always belong to an application, like elements and conditions.

Those are implicitly forced to belong to an application anyway.

The other generic macros cannot be moved ofc

Regarding the registration of the core things, do you think it would make sense to move those to the Register fct of the core?

Comment (roigcarlo) [reply kratos/includes/define.h:707]: The idea is to move everything to the registry and remove KratosComponents, eventually...

Comment (roigcarlo) [reply kratos/includes/define.h:707]: If you preffer, instead of using `mApplicationName` as the key I can use the mangled name of the class executing the macro (as a matter of fact, that was my first approach). That way no compilation errors should ever arise, but it makes deregistering things much more cumbersome as you have to ensure that the register and deregister function exists in an isolated context and its consistent across the application :S

File: applications/IgaApplication/iga_application.cpp (hunk 4 above)
Comment (roigcarlo) [reply applications/IgaApplication/iga_application.cpp:67]: Aside from the fact that IGA has a lot of dependencies in the tests, all fine:

```console
Running IgaApplication tests
/home/roigcarlo/Kratos/bin/Debug/applications//IgaApplication/tests/test_IgaApplication.py
....................................
----------------------------------------------------------------------
Ran 36 tests in 41.760s

OK
Completed IgaApplication tests.
```

+ 

```c++
Test results summary:
  KratosCore                     : OK
  CoSimulationApplication        : OK
  FluidDynamicsApplication       : OK
  IgaApplication                 : OK
  LinearSolversApplication       : OK
  StructuralMechanicsApplication : OK
  TrilinosApplication            : OK
  cpp                            : OK
```

File: kratos/includes/define.h (hunk 6 above)
Comment (philbucher) [reply kratos/includes/define.h:707]: Hm I see

Well if you have a global plan then no complaints from my side 😄

File: kratos/includes/kratos_components.h [hunk 10]
```diff
@@ -93,6 +96,25 @@ class KRATOS_API(KRATOS_CORE) KratosComponents
...
+     * @brief Adds the source of a component to the list of loaded components.
+     * @param rSrcName the name of the source who added the component
+     * @param rCmpName the name of the component to add
+     * @throws std::runtime_error if the same object name was already registered in the same origin
+     */
+    static void AddSource(const std::string& rSrcName, const std::string& rCmpName)
+    {
+        // Create the container if its the first component source added
+        if (msComponentsSources.find(rSrcName) == msComponentsSources.end()) {
+            msComponentsSources[rSrcName] = std::unordered_set<std::string>();
```
Comment (philbucher) [inline kratos/includes/kratos_components.h]: Hm still this lookup is done 3 times

File: kratos/includes/kratos_components.h [hunk 11]
```diff
@@ -93,6 +96,25 @@ class KRATOS_API(KRATOS_CORE) KratosComponents
...
+        }
+
+        // Check if a different object was already registered with this name
+        auto it_comp =  msComponentsSources[rSrcName].find(rCmpName);
```
Comment (philbucher) [inline kratos/includes/kratos_components.h]: Number 2

File: kratos/includes/kratos_components.h [hunk 12]
```diff
@@ -93,6 +96,25 @@ class KRATOS_API(KRATOS_CORE) KratosComponents
...
+        KRATOS_ERROR_IF(it_comp != msComponentsSources[rSrcName].end()) << "An object with name \"" << rCmpName << " has already been registered for " << rSrcName << std::endl;
```
Comment (philbucher) [inline kratos/includes/kratos_components.h]: Number 3

**Review by philbucher — DISMISSED** [review]
Minor performance related comment

File: kratos/includes/kratos_components.h (hunk 12 above)
Comment (philbucher) [inline kratos/includes/kratos_components.h]:
```suggestion
        KRATOS_ERROR_IF(it_comp != msComponentsSources[rSrcName].end()) << "An object with name \"" << rCmpName << "\" has already been registered for " << rSrcName << std::endl;
```

**Review by philbucher — APPROVED** [review]
great

**Comment by roigcarlo** [comment]
Ok no lookups.

Also after fixing the `if` condition I realized there are a couple of apps that will make it throw until the gtest PR enters, so I commented it for now. Nothing wring with what they do, is just not compatible and not gonna bother to fix it as the tests PR will fix it automatically. (I.e. `Chsarp` and `Geomechancis` because they explicitly call the `StructuralMechancis` kernel register)

**Comment by philbucher** [comment]
Hm those register calls are wrong

Can you make a separate PR removing them?

**Comment by roigcarlo** [comment]
Nope, that's the problem, technically speaking, they are ok. Gonna try to explain the whole situation, but will be a little long:

Historically we have omitted the possibility of Kratos being called from anywhere different from python. Since the test are running through python anyway that was not a problem, but we end up in a situation where Python initializes the kernel and every time it imports a library are registers it in the kernel.  This kernel and applications are never destroyed and the destruction code does not handle what happens with the components they registered. 

This is what I am going to call **Scenario A**, and it what is happening in the current master.

At some point, some applications arise where the life cycle of the kernel is directly controlled by the apps. To the date this apps are the CSharpApplication and the GeoMechancis ( probably more that I am not aware ). A normal execution of those applications passes by referencing the current instance of the kernel and then managing the applications that are registered. For example for the GeoMechanics:

```c++
    class GeoWorkflow: 
    
GeoWorkflow::GeoWorkflow() {
    ....
    if (!mKernel.IsImported("GeoMechanicsApplication")) {
        KRATOS_INFO("KratosExecute") << "Importing GeoMechanicsApplication" << std::endl;
        mpGeoApp = Kratos::make_shared<KratosGeoMechanicsApplication>();
        mKernel.ImportApplication(mpGeoApp);
    }
}
```

This means that someone can create its main, initialize the kratos kernel, call the `GeoWorkflow` and then import its application without major problems. The important detail here is that this wrapper or workflow has a reference to a given kernel objects, and since the kernel is static, at all effects doing this takes "ownership" over the kernel in terms of c++ object lifetime. 

Similarly in the CSharpApplication this is done for the structural but is never checked if the application was already registered because its assumed the wrapper launches with the executable, not per use case ( this I can fix it )

Lets call this **Scenario B**

Now, both A & B work under the assumption that you will always have a kernel object that contains all the information that you need, and this is true as long as we check if things were loaded before and never unload anything from the kernel, because either the C++ main or python are the owners. But it starts to fail the moment we want to destroy or recreate the kernel or change the set of components that are registered at any time. Some use cases:

## 1 - Python / workflow / workflow
### Without deregister
1 - Python is initialized, Kernel and apps are loaded
2 - First workflow takes ownership over the kernel and tries to register other apps ( can do because checks if apps are repeated )
3 - Test ends and destroys the kernel object. Does nothing because the kernel destroyer does nothing and app was not created because it existed in the kernel.
4 - Second workflow takes ownership over the kernel and tries to register other apps ( can do because kernel is still valid and checks if apps are repeated )
PASS

### With deregister
1 - Python is initialized, Kernel and apps are loaded
2 - First workflow takes ownership over the kernel and tries to register other apps ( can do because checks if apps are repeated )
3 - Test ends and destroys the kernel object. Kernel unregistered is called, all objects registered by the kernel are unloaded.
4 - Second workflow takes ownership over the kernel and tries to register other apps. Can do it because the kernel still have info about the app being registered, but now it will lack all the components registered from the core 
CRASH

## 2 - Binary / workflow / workflow
### Without deregister
1 - Exe is initialized, Kernel is created (no apps loaded)
2 - First workflow takes ownership over the kernel and tries to register other apps ( can do because none is registered )
3 - Test ends and destroys the kernel object. Does nothing because the kernel destroyer does nothing. App reference is destroyed but does nothing because the destroyer is empty
4 - Second workflow takes ownership over the kernel and tries to register other apps ( can do because kernel is still valid and will see app as registered )
PASS

### With deregister
1 - Exe is initialized, Kernel is created (no apps loaded)
2 - First workflow takes ownership over the kernel and tries to register other apps ( can do because none is registered )
3 - Test ends and destroys the kernel object. Kernel unregistered is called, all objects registered by the kernel are unloaded. App unregistered is called, all objects registered by the app are unloaded and the app is not in the app list anymore.
4 - Second workflow takes ownership over the kernel and tries to register other apps. Can do it because the app was correctly unloaded, but now it will lack all the components registered from the core 
CRASH

## 3 - Binary with isolation / workflow / workflow
Here I assume that kernel object is something that can be created and destroyed across an execution.
### Without deregister
1 - Exe is initialized.
2 - First workflow takes creates a kernel and tries to register other apps ( can do because none is registered and kernel did not exist )
3 - Test ends and destroys the kernel object. Does nothing because the kernel destroyer does nothing. App reference is destroyed but does nothing because the destroyer is empty
4 - Second workflow creates a kernel and crashes because all components were already registered.
CRASH

### With deregister
1 - Exe is initialized, Kernel is created (no apps loaded)
2 - First workflow creates a kernel and tries to register other apps ( can do because none is registered and kernel did not exist )
3 - Test ends and destroys the kernel object. Kernel unregistered is called, all objects registered by the kernel are unloaded. App unregistered is called, all objects registered by the app are unloaded and the app is not in the app list anymore.
4 - Second workflow takes ownership over the kernel and tries to register other apps. Can do it because kernel was correctly unloaded and apps were correctly unloaded as well. No repeated components or app appearing in the list of loaded apps.
PASS

Use case 3 is basically what will happen when we merge the GTest PR. Use case 1 makes A pass but B no because of how things are registered, but the problem is not per se B being wrong, is B because python cannot handle the fact that the kernel object is destroyed because we do not have any way to recreate it if unloaded correctly. Similarly Use case 2 acts as 1 because we are treating the main of the binary as the python script, with the salvo that is infinitely easier to fill, as we can create and destroy the kernel object ( which is actually case 3) in the main.

TL;DR:

Kratos does not handle the kernel correctly. Fixing it and having all checks in place will make some existing tests to fail because they relay on wrong behavior (the ones fixed in the GTest PR) or because they relay on the kernel being clean ant the start ( the ones failing in this PR if I activate the check). Both tests cannot be fixed at once, and the problem with the later ones is in the current tests execution, no in the tests or the workflows themselves.

**Comment by roigcarlo** [comment]
> Hm those register calls are wrong
> 
> Can you make a separate PR removing them?

Well as a matter of fact, all of them will work again in #11506. But at least now we have an explanation of what is happening somewhere :P

**Review by pooyan-dadvand — COMMENTED** [review]
I tend to agree with @philbucher that the solution is hacky but there are no clean solution without changing the interface.

Nevertheless, I would recommend to use the registry for `applications` indexing so we can reuse the mechanism when we migrate everything to the registry.

**Comment by pooyan-dadvand** [comment]
Also I would suggest to use a normal Application function instead of macro to set the application name in the mApplicationName

**Comment by roigcarlo** [comment]
New version in #12306

**Comment by loumalouomega** [comment]
> New version in #12306

Should you close this one?

**Comment by roigcarlo** [comment]
Yup, but lets wait, i have changes here I want to peek and its better If I can browse the branch in the nav :)
